File: frb_example/pure_dart/lib/src/rust/api/enumeration.dart
// ignore_for_file: invalid_use_of_internal_member, unused_import

import '../frb_generated.dart';
import 'misc_example.dart';
import 'package:flutter_rust_bridge/flutter_rust_bridge_for_generated.dart';
import 'package:freezed_annotation/freezed_annotation.dart' hide protected;
part 'enumeration.freezed.dart';

Future<EnumSimpleTwinNormal> funcEnumSimpleTwinNormal(
        {required EnumSimpleTwinNormal arg, dynamic hint}) =>
    RustLib.instance.api.funcEnumSimpleTwinNormal(arg: arg, hint: hint);

Future<EnumWithItemMixedTwinNormal> funcEnumWithItemMixedTwinNormal(
        {required EnumWithItemMixedTwinNormal arg, dynamic hint}) =>
    RustLib.instance.api.funcEnumWithItemMixedTwinNormal(arg: arg, hint: hint);

Future<EnumWithItemTupleTwinNormal> funcEnumWithItemTupleTwinNormal(
        {required EnumWithItemTupleTwinNormal arg, dynamic hint}) =>
    RustLib.instance.api.funcEnumWithItemTupleTwinNormal(arg: arg, hint: hint);

Future<EnumWithItemStructTwinNormal> funcEnumWithItemStructTwinNormal(
        {required EnumWithItemStructTwinNormal arg, dynamic hint}) =>
    RustLib.instance.api.funcEnumWithItemStructTwinNormal(arg: arg, hint: hint);

Future<Uint8List> printNote({required Note note, dynamic hint}) =>
    RustLib.instance.api.printNote(note: note, hint: hint);

Future<Weekdays?> handleReturnEnum({required String input, dynamic hint}) =>
    RustLib.instance.api.handleReturnEnum(input: input, hint: hint);

Future<Weekdays> handleEnumParameter(
        {required Weekdays weekday, dynamic hint}) =>
    RustLib.instance.api.handleEnumParameter(weekday: weekday, hint: hint);

Future<Measure?> multiplyByTen({required Measure measure, dynamic hint}) =>
    RustLib.instance.api.multiplyByTen(measure: measure, hint: hint);

Future<KitchenSink> handleEnumStruct(
        {required KitchenSink val, dynamic hint}) =>
    RustLib.instance.api.handleEnumStruct(val: val, hint: hint);

@freezed
sealed class Distance with _$Distance {
  const factory Distance.unknown() = Distance_Unknown;
  const factory Distance.map(
    double field0,
  ) = Distance_Map;
}

enum EnumSimpleTwinNormal {
  a,
  b,
}

@freezed
sealed class EnumWithItemMixedTwinNormal with _$EnumWithItemMixedTwinNormal {
  const factory EnumWithItemMixedTwinNormal.a() = EnumWithItemMixedTwinNormal_A;
  const factory EnumWithItemMixedTwinNormal.b(
    Uint8List field0,
  ) = EnumWithItemMixedTwinNormal_B;
  const factory EnumWithItemMixedTwinNormal.c({
    required String cField,
  }) = EnumWithItemMixedTwinNormal_C;
}

@freezed
sealed class EnumWithItemStructTwinNormal with _$EnumWithItemStructTwinNormal {
  const factory EnumWithItemStructTwinNormal.a({
    required Uint8List aField,
  }) = EnumWithItemStructTwinNormal_A;
  const factory EnumWithItemStructTwinNormal.b({
    required Int32List bField,
  }) = EnumWithItemStructTwinNormal_B;
}

@freezed
sealed class EnumWithItemTupleTwinNormal with _$EnumWithItemTupleTwinNormal {
  const factory EnumWithItemTupleTwinNormal.a(
    Uint8List field0,
  ) = EnumWithItemTupleTwinNormal_A;
  const factory EnumWithItemTupleTwinNormal.b(
    Int32List field0,
  ) = EnumWithItemTupleTwinNormal_B;
}

@freezed
sealed class KitchenSink with _$KitchenSink {
  /// Comment on variant
  const factory KitchenSink.empty() = KitchenSink_Empty;
  const factory KitchenSink.primitives({
    /// Dart field comment
    @Default(-1) int int32,
    required double float64,
    required bool boolean,
  }) = KitchenSink_Primitives;
  const factory KitchenSink.nested(
    int field0, [
    @Default(KitchenSink.empty()) KitchenSink field1,
  ]) = KitchenSink_Nested;
  const factory KitchenSink.optional([
    /// Comment on anonymous field
    @Default(-1) int? field0,
    int? field1,
  ]) = KitchenSink_Optional;
  const factory KitchenSink.buffer(
    Uint8List field0,
  ) = KitchenSink_Buffer;
  const factory KitchenSink.enums([
    @Default(Weekdays.sunday) Weekdays field0,
  ]) = KitchenSink_Enums;
}

@freezed
sealed class Measure with _$Measure {
  const factory Measure.speed(
    Speed field0,
  ) = Measure_Speed;
  const factory Measure.distance(
    Distance field0,
  ) = Measure_Distance;
}

class Note {
  final Weekdays day;
  final String body;

  const Note({
    this.day = Weekdays.sunday,
    required this.body,
  });

  @override
  int get hashCode => day.hashCode ^ body.hashCode;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Note &&
          runtimeType == other.runtimeType &&
          day == other.day &&
          body == other.body;
}

@freezed
sealed class Speed with _$Speed {
  const factory Speed.unknown() = Speed_Unknown;
  const factory Speed.gps(
    double field0,
  ) = Speed_GPS;
}

File: frb_example/pure_dart/rust/src/api/enumeration.rs
use crate::api::misc_example::WeekdaysTwinNormal;
use crate::frb_generated::StreamSink;
use flutter_rust_bridge::frb;
use log::info;
use serde::{Deserialize, Serialize};

pub enum EnumSimpleTwinNormal {
    A,
    B,
}

pub fn func_enum_simple_twin_normal(arg: EnumSimpleTwinNormal) -> EnumSimpleTwinNormal {
    arg
}

pub enum EnumWithItemMixedTwinNormal {
    A,
    B(Vec<u8>),
    C { c_field: String },
}

pub fn func_enum_with_item_mixed_twin_normal(
    arg: EnumWithItemMixedTwinNormal,
) -> EnumWithItemMixedTwinNormal {
    arg
}

pub enum EnumWithItemTupleTwinNormal {
    A(Vec<u8>),
    B(i32),
}

pub fn func_enum_with_item_tuple_twin_normal(
    arg: EnumWithItemTupleTwinNormal,
) -> EnumWithItemTupleTwinNormal {
    arg
}

pub enum EnumWithItemStructTwinNormal {
    A { a_field: Vec<u8> },
    B { b_field: Vec<i32> },
}

pub fn func_enum_with_item_struct_twin_normal(
    arg: EnumWithItemStructTwinNormal,
) -> EnumWithItemStructTwinNormal {
    arg
}

// #1674
pub enum EnumWithDiscriminantTwinNormal {
    OneHundred = 100,
    Fifty = 50,
}

pub fn func_enum_with_discriminant_twin_normal(
    arg: EnumWithDiscriminantTwinNormal,
) -> EnumWithDiscriminantTwinNormal {
    arg
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ChangeTwinNormal<T> {
    Created { data: T },
    Updated { id: String, data: T },
    Deleted { id: String },
}

pub type ChangeStringTwinNormal = ChangeTwinNormal<String>;

pub fn func_change_twin_normal(arg: ChangeStringTwinNormal) -> ChangeStringTwinNormal {
    arg
}

pub type ChangeMapTwinNormal = ChangeTwinNormal<std::collections::HashMap<String, String>>;

pub fn func_change_map_twin_normal(arg: ChangeMapTwinNormal) -> ChangeMapTwinNormal {
    arg
}

// Test nested generic enum inside StreamSink - this tests auto-naming for nested generics
pub type ChangeMapSinkTwinNormal = StreamSink<ChangeTwinNormal<std::collections::HashMap<String, String>>>;

pub fn func_change_map_sink_twin_normal(arg: ChangeMapSinkTwinNormal) -> ChangeMapSinkTwinNormal {
    arg
}

// Test case for bug: generic enum with #[frb(ignore)] and concrete type aliases with #[frb(non_opaque)]
// The template enum is ignored, but concrete instances through type aliases should generate Dart code
#[derive(Debug, Clone, Serialize, Deserialize)]
#[flutter_rust_bridge::frb(ignore)]
pub enum ChangeIgnoredTwinNormal<T> {
    Created { data: T },
    Updated { id: String, data: T },
    Deleted { id: String },
}

#[frb(non_opaque)]
pub type StringChangeTwinNormal = ChangeIgnoredTwinNormal<String>;

#[frb(non_opaque)]
pub type MapChangeTwinNormal = ChangeIgnoredTwinNormal<std::collections::HashMap<String, String>>;

pub fn func_block_change_twin_normal(arg: StringChangeTwinNormal) -> StringChangeTwinNormal {
    arg
}

pub fn func_map_change_twin_normal(arg: MapChangeTwinNormal) -> MapChangeTwinNormal {
    arg
}

pub struct BatchTwinNormal<T> {
    items: Vec<T>,
}

#[frb(non_opaque)]
pub type BatchTwinNormalChangeMapTwinNormal = BatchTwinNormal<ChangeIgnoredTwinNormal<std::collections::HashMap<String, String>>>;

pub fn func_batch_change_map_twin_normal(arg: BatchTwinNormalChangeMapTwinNormal) -> BatchTwinNormalChangeMapTwinNormal {
    arg
}

#[frb(non_opaque)]
pub struct AnotherBatchTwinNormal<T> {
    items: T,
}


#[frb(non_opaque)]
pub type AnotherBatchTwinNormalChangeMapTwinNormal = AnotherBatchTwinNormal<ChangeIgnoredTwinNormal<std::collections::HashMap<String, String>>>;

pub fn func_another_batch_twin_normal_change_map_twin_normal(arg: AnotherBatchTwinNormalChangeMapTwinNormal) -> AnotherBatchTwinNormalChangeMapTwinNormal {
    arg
}

#[frb]
#[derive(Debug)]
pub struct NoteTwinNormal {
    #[frb(default = "WeekdaysTwinNormal.Sunday")]
    pub day: Box<WeekdaysTwinNormal>,
    pub body: String,
}

pub fn print_note_twin_normal(note: NoteTwinNormal) -> Vec<u8> {
    info!("{:#?}", note);
    vec![1, 2, 3]
}

pub fn handle_return_enum_twin_normal(input: String) -> Option<WeekdaysTwinNormal> {
    match input.as_str() {
        "Monday" => Some(WeekdaysTwinNormal::Monday),
        "Tuesday" => Some(WeekdaysTwinNormal::Tuesday),
        "Wednesday" => Some(WeekdaysTwinNormal::Wednesday),
        "Thursday" => Some(WeekdaysTwinNormal::Thursday),
        "Friday" => Some(WeekdaysTwinNormal::Friday),
        "Saturday" => Some(WeekdaysTwinNormal::Saturday),
        "Sunday" => Some(WeekdaysTwinNormal::Sunday),
        _ => None,
    }
}

pub fn handle_enum_parameter_twin_normal(weekday: WeekdaysTwinNormal) -> WeekdaysTwinNormal {
    info!("The weekday is {:?}", weekday);
    weekday
}

#[derive(Debug, Clone)]
pub enum SpeedTwinNormal {
    Unknown,
    GPS(f64),
}

#[derive(Debug, Clone)]
pub enum DistanceTwinNormal {
    Unknown,
    Map(f64),
}

#[derive(Debug, Clone)]
pub enum MeasureTwinNormal {
    Speed(Box<SpeedTwinNormal>),
    Distance(Box<DistanceTwinNormal>),
}

pub fn multiply_by_ten_twin_normal(measure: MeasureTwinNormal) -> Option<MeasureTwinNormal> {
    match measure {
        MeasureTwinNormal::Speed(b) => match *b {
            SpeedTwinNormal::GPS(v) => Some(MeasureTwinNormal::Speed(Box::new(
                SpeedTwinNormal::GPS(v * 10.),
            ))),
            SpeedTwinNormal::Unknown => None,
        },
        MeasureTwinNormal::Distance(b) => match *b {
            DistanceTwinNormal::Map(v) => Some(MeasureTwinNormal::Distance(Box::new(
                DistanceTwinNormal::Map(v * 10.),
            ))),
            DistanceTwinNormal::Unknown => None,
        },
    }
}

#[frb]
#[derive(Debug)]
pub enum KitchenSinkTwinNormal {
    /// Comment on variant
    Empty,
    #[frb(unimpl_variant_attr)]
    Primitives {
        /// Dart field comment
        #[frb(default = -1)]
        int32: i32,
        float64: f64,
        boolean: bool,
    },
    Nested(
        i32,
        #[frb(default = "KitchenSinkTwinNormal.empty()")] Box<KitchenSinkTwinNormal>,
    ),
    Optional(
        /// Comment on anonymous field
        #[frb(default = -1)]
        Option<i32>,
        Option<i32>,
    ),
    Buffer(Vec<u8>),
    Enums(#[frb(default = "WeekdaysTwinNormal.Sunday")] WeekdaysTwinNormal),
}

pub fn handle_enum_struct_twin_normal(val: KitchenSinkTwinNormal) -> KitchenSinkTwinNormal {
    use KitchenSinkTwinNormal::*;
    use WeekdaysTwinNormal::*;
    let inc = |x| x + 1;
    match val {
        Primitives {
            int32,
            float64,
            boolean,
        } => Primitives {
            int32: int32 + 1,
            float64: float64 + 1.,
            boolean: !boolean,
        },
        Nested(val, _nested) => Nested(inc(val), Box::new(KitchenSinkTwinNormal::Empty)),
        Optional(a, b) => Optional(a.map(inc), b.map(inc)),
        Buffer(mut buf) => {
            buf.push(1);
            Buffer(buf)
        }
        Enums(day) => Enums(match day {
            Monday => Tuesday,
            Tuesday => Wednesday,
            Wednesday => Thursday,
            Thursday => Friday,
            Friday => Saturday,
            Saturday => Sunday,
            Sunday => Monday,
        }),
        _ => val,
    }
}
